Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,102 @@
+
+private typealias VersionFile = [String: [String: AnyObject]]
+
+public func createVersionFileForDependency(dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, buildProductURLs: [NSURL], rootDirectoryURL: NSURL) -> Bool {
+	var cachedPlatformFrameworks: [String: [[String: String]]] = [:]
+	
+	for url in buildProductURLs {
+		guard let platformName = url.URLByDeletingLastPathComponent?.lastPathComponent else { return false }
+		guard let frameworkName = url.URLByDeletingPathExtension?.lastPathComponent else { return false }
+		
+		let frameworkURL = url.appendingPathComponent(frameworkName, isDirectory: false)
+		guard NSFileManager.defaultManager().fileExistsAtPath(frameworkURL.path!) else { return false }
+		guard let sha1 = sha1ForFileAtURL(frameworkURL) else { return false }
+		
+		let frameworkInfo = [frameworkNameKeyName: frameworkName, frameworkSha1KeyName: sha1]
+		
+		var frameworks = cachedPlatformFrameworks[platformName] ?? []
+		frameworks.append(frameworkInfo)
+		cachedPlatformFrameworks[platformName] = frameworks
+	}
+	
+	var versionInfo: VersionFile = [:]
+	
+	let platformSet = platforms.isEmpty ? Set(Platform.supportedPlatforms) : platforms
+	for platform in platformSet {
+		let cachedFrameworks: [[String: String]] = cachedPlatformFrameworks[platform.rawValue] ?? []
+		versionInfo[platform.rawValue] = [commitishKeyName: dependency.version.commitish, cachedFrameworksKeyName: cachedFrameworks]
+	}
+	
+	let rootBinariesURL = rootDirectoryURL.appendingPathComponent(CarthageBinariesFolderPath, isDirectory: true).URLByResolvingSymlinksInPath!
+	let versionFileURL = rootBinariesURL.appendingPathComponent(".\(dependency.project.name).version")
+	
+	if let oldVersionFile: VersionFile = readVersionFileAtURL(versionFileURL) {
+		for platformName in oldVersionFile.keys {
+			if versionInfo[platformName] == nil {
+				versionInfo[platformName] = oldVersionFile[platformName]
+			}
+		}
+	}
+
+	do {
+		let jsonData = try NSJSONSerialization.dataWithJSONObject(versionInfo, options: .PrettyPrinted)
+		try jsonData.writeToURL(versionFileURL, options: .DataWritingAtomic)
+	}
+	catch { return false }
+	
+	return true
+}
+
+public func versionFileMatchesDependency(dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, rootDirectoryURL: NSURL) -> Bool {
+	let rootBinariesURL = rootDirectoryURL.appendingPathComponent(CarthageBinariesFolderPath, isDirectory: true).URLByResolvingSymlinksInPath!
+	let versionFileURL = rootBinariesURL.appendingPathComponent(".\(dependency.project.name).version")
+	guard let versionInfo = readVersionFileAtURL(versionFileURL) else { return false }
+	
+	let platformNames: [String] = !platforms.isEmpty ? platforms.map { $0.rawValue } : Array(versionInfo.keys)
+	guard platformNames.count > 0 else { return false }
+	
+	for platformName in platformNames {
+		guard let platformInfo: [String: AnyObject] = versionInfo[platformName] else { return false }
```

Comment:
This versioning information seems like it deserves a proper model type, rather than just dictionary lookups.